web-api/src/storage/elastic/filter.rs

```diff
@@ -12,36 +12,81 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+use itertools::Itertools;
 use serde_json::{json, Value};
 use xayn_web_api_shared::serde::JsonObject;
 
 use crate::{
-    personalization::filter::{CompareOp, Filter},
+    personalization::filter::{CombineOp, CompareOp, Filter},
     storage::KnnSearchParams,
 };
 
-fn extend_filter(filter: &mut JsonObject, clause: &Filter) {
-    let clause = match clause {
+fn extend_value(filter: &mut JsonObject, occurence: &'static str, mut clause: Value) {
+    if let Some(filter) = filter.get_mut(occurence) {
+        match (&mut *filter, &mut clause) {
+            (Value::Array(filter), Value::Array(clause)) => filter.append(clause),
+            (Value::Array(filter), Value::Object(_)) => filter.push(clause),
+            (Value::Object(_), Value::Array(clauses)) => {
+                clauses.insert(0, filter.take());
+                *filter = clause;
+            }
+            (Value::Object(_), Value::Object(_)) => {
+                *filter = json!([filter.take(), clause]);
+            }
+            _ => unreachable!(/* filter and clause are array or object */),
+        }
+    } else {
+        filter.insert(occurence.to_string(), clause);
+    }
+}
+
+fn extend_filter(filter: &mut JsonObject, clause: &Filter, is_not_root: Option<&'static str>) {
```

Comment:
I will approve it for now but I feel the readability and maintainability of this code is pretty bad.

Some ideas:

- maybe we probably should not special case "is root"/"is not root"
- avoid or change how we do some of the micro optimizations like the `"should": [{...}]` as `"should": {...}` part

Reply:
> maybe we probably should not special case "is root"/"is not root"

i agree with that, but it requires further refactoring in the es modules, because we would also need to apply the `{"bool": {"filter": object}}` structure to the root level, but currently we carry only the `object` thingy around. hence i didn't make these changes yet, but we definitely should follow up on this.

> avoid or change how we do some of the micro optimizations like the "should": [{...}] as "should": {...} part

i was wondering how es handles that internally, does it make a difference performance wise? or is it handled as array of objects anyways?

Reply:
i simplified the code such that all clauses must be arrays now, this removes a lot of the array/object handling.